plugins/tedge_apama_plugin/src/main.rs

```diff
@@ -50,123 +49,217 @@ const TEDGE_APAMA_PROJECT_DIR: &str = "/etc/tedge/apama/project";
+    }
+
+    Ok(())
+}
+
+fn get_installed_monitors() -> Result<Vec<String>, InternalError> {
+    // Run `engine_inspect -m -r` command to list all monitors in raw format
+    let output = Command::new(APAMA_ENV_EXE)
+        .arg(ENGINE_INSPECT_CMD)
+        .arg("-m")
+        .arg("-r")
+        .stdin(Stdio::null())
+        .output()
+        .map_err(|err| InternalError::exec_error(ENGINE_INSPECT_CMD, err))?;
+    let output = String::from_utf8(output.stdout)?;
+
+    // The output contains monitor names and their instance counts separated by a space as follows:
+    // ```
+    // TedgeDemoMonitor 1
+    // TedgeTestMonitor 1
+    // ```
+    let mon_files = output
+        .lines()
+        // The first line of the output could "WARNING: JAVA_HOME not set" which is filtered out
+        .filter(|line| !line.starts_with("WARNING:"))
+        // The counts are filtered out too
+        .filter_map(|line| line.split_whitespace().next())
+        .map(|line| line.into())
+        .collect();
+
+    Ok(mon_files)
+}
+
+fn install_or_update_monitor(mon_name: &str, mon_file_path: &str) -> Result<(), InternalError> {
+    let installed_monitors = get_installed_monitors()?;
+    // If an existing monitor needs to be updated, older version needs to be removed first before installing the new one
+    if installed_monitors.contains(&mon_name.to_string()) {
+        remove_monitor(mon_name)?;
+        install_monitor(mon_file_path)
+    } else {
+        install_monitor(mon_file_path)
+    }
+}
+
+fn install_monitor(mon_file_path: &str) -> Result<(), InternalError> {
+    run_cmd(
+        APAMA_ENV_EXE,
+        format!("{} {}", ENGINE_INJECT_CMD, mon_file_path).as_str(),
+    )
+}
+
+fn remove_monitor(mon_name: &str) -> Result<(), InternalError> {
+    run_cmd(
+        APAMA_ENV_EXE,
+        format!("{} {}", ENGINE_DELETE_CMD, mon_name).as_str(),
+    )
+}
+
+fn run_cmd(cmd: &str, args: &str) -> Result<(), InternalError> {
     let args: Vec<&str> = args.split_whitespace().collect();
-    let status = Command::new(cmd)
+    let exit_status = Command::new(cmd)
         .args(args)
         .stdin(Stdio::null())
         .status()
         .map_err(|err| InternalError::exec_error(cmd, err))?;
-    Ok(status)
+
+    if exit_status.success() {
+        Ok(())
+    } else {
+        Err(InternalError::ExecFailure {
+            cmd: cmd.into(),
+            exit_status,
+        })
+    }
 }
 
 fn main() {
     // On usage error, the process exits with a status code of 1
     let apama = ApamaCli::from_args();
 
     match run(apama.operation) {
-        Ok(status) if status.success() => {
+        Ok(()) => {
             std::process::exit(0);
         }
 
-        Ok(status) => {
-            if status.code().is_some() {
-                std::process::exit(2);
-            } else {
-                eprintln!("Interrupted by a signal!");
-                std::process::exit(4);
-            }
-        }
-
```

Comment:
This needs to be reverted.

Reply:
Even the earlier version was flawed as per the contract that @makr11st has pasted above. So, instead of reverting, I've just fixed the logic so that it's aligned with the spec. It's a much simpler contract now. If there are any failures during the plugin execution, either in the rust logic or some external command execution, we fail with exit code 2 (we were failing with exit code 5 earlier), except for usage error which will still fail with exit code 1.